File: design-patterns/go/builder/vehicle/car_builder.go
// car_builder.go

package main

import "fmt"

type CarBuilder struct {
	wheel    int
	engine   int
	seat     int
	interior bool
	door     int
}

func NewCarBuilder() (carBuilder *CarBuilder) {
	carBuilder = &CarBuilder{}
	return
}

func (carBuilder *CarBuilder) AddDoor(noOfDoor int) {
	fmt.Printf("Add %d door\n", noOfDoor)
	carBuilder.door += noOfDoor
}

func (carBuilder *CarBuilder) AddEngine(noOfEngine int) {
	fmt.Printf("Add %d engine\n", noOfEngine)
	carBuilder.engine += noOfEngine
}

func (carBuilder *CarBuilder) AddInterior() {
	fmt.Println("Add interior")
	carBuilder.interior = true
}

func (carBuilder *CarBuilder) AddSeat(noOfSeat int) {
	fmt.Printf("Add %d seat\n", noOfSeat)
	carBuilder.seat = noOfSeat
}

func (carBuilder *CarBuilder) AddWheel(noOfWheel int) {
	fmt.Printf("Add %v wheel\n", noOfWheel)
	carBuilder.wheel += noOfWheel
}

func (carBuilder *CarBuilder) AddWing(noOfWings int) {
	panic("Can not add wings")
}

func (carBuilder *CarBuilder) Build() *Car {
	car := NewCar(carBuilder.wheel, carBuilder.engine, carBuilder.seat, carBuilder.door, carBuilder.interior)
	return car
}

File: design-patterns/ts/builder/vehicle/car-builder.ts
// car-builder.ts

import Car from "./car";
import VehicleBuilder from "./vehicle-builder";


class CarBuilder implements VehicleBuilder {
    private wheel: number = 0;
    private engine: number = 0;
    private seat!: number;
    private interior!: boolean;
    private door: number = 0;

    addWheel(noOfWheel: number) {
        console.log("Add " + noOfWheel + " wheels");
        this.wheel += noOfWheel;
    }

    addEngine(noOfEngine: number) {
        console.log("Add " + noOfEngine + " engine");
        this.engine += noOfEngine;
    }

    addSeat(noOfSeat: number) {
        console.log("Add " + noOfSeat + " Seat");
        this.seat = noOfSeat;
    }

    addInterior() {
        console.log("Add interior");
        this.interior = true;
    }

    addDoor(noOfDoor: number) {
        console.log("Add " + noOfDoor + " door");
        this.door += noOfDoor;
    }

    addWing(noOfWings: number) {
        throw new Error("Can not add wings");
    }

    build(): Car {
        const car: Car = new Car(this.wheel, this.engine, this.seat, this.door, this.interior);

        return car;
    }
}

export default CarBuilder;